w11b01/test/pgdp/pengusurvivors/ResidualGraphTest.java

```diff
@@ -0,0 +1,178 @@
+
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class ResidualGraphTest {
+
+
+    @ParameterizedTest
+    @MethodSource("getResidualGraphArguments")
+    void shouldGenerateCorrectResidualGraph(String source, String expectedResidual) {
+        var sourceGraph = GraphGenerator.generateGraph(source).graph();
+        var residualGraph = GraphGenerator.generateGraph(expectedResidual).graph();
+        sourceGraph.generateResidualGraph();
+        assertTrue(GraphUtils.residualEquals(sourceGraph, residualGraph));
+    }
+
+    static Stream<Arguments> getResidualGraphArguments() {
+        return Stream.of(
+                Arguments.of(
+                        """
+                                a -> b (4/10);
+                                b -> c (2/6);
+                                """,
+                        """
+                                a -> b (0/6);
+                                b -> a (0/4);
+                                b -> c (0/4);
+                                c -> b (0/2);
+                                """
+                )
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("findPathArguments")
+    void findPath(String verticeString, String source, String sink, List<List<String>> expectedPaths) {
+        var graphResult = GraphGenerator.generateGraph(
+                verticeString,
+                source,
+                sink
+        );
+        var graph = graphResult.graph();
+        GraphUtils.printGraph(graph);
+        var vertices = graphResult.vertices();
+        graph.generateResidualGraph();
+        var path = graph.findPathInResidual();
+        var matchingPaths = 0;
+        if (expectedPaths.isEmpty()) {
+            if (path != null)
+                throw new AssertionError("Expected no path, but found " + GraphUtils.stringifyPath(path));
+            return;
+        }
+        pathloop:
+        for (var expectedPath : expectedPaths) {
+            if (path == null || expectedPath.size() != path.size()) continue;
+            for (int i = 0; i < path.size(); i++) {
+                var expectedVertex = vertices.get(expectedPath.get(i));
+                var actualVertex = path.get(i);
+                if (expectedVertex != actualVertex) continue pathloop;
+            }
+            matchingPaths++;
+        }
+        assertEquals(1, matchingPaths, "No matching path was found. Your path was: " + (path == null ? "None found" : GraphUtils.stringifyPath(path)));
+
+    }
+
+    static Stream<Arguments> findPathArguments() {
+        return Stream.of(
+                Arguments.of("""
+                        s -> a (1/1);
+                        s -> b (0/1);
+                        a -> c (1/1);
+                        b -> c (0/1);
+                        """, "s", "c", List.of(List.of("s", "b", "c"))),
```

Comment:
`s -> a -> c` should also be valid, right?

```suggestion
                        """, "s", "c", List.of(List.of("s", "b", "c"), List.of("s", "a", "c"))),
```